tiledb/sm/query/ast/query_ast.cc

```diff
@@ -0,0 +1,143 @@
+/**
+ * @file query_ast.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2021 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Implements Query AST classes.
+ */
+
+#include "query_ast.h"
+
+namespace tiledb {
+namespace sm {
+
+tdb_unique_ptr<ASTNode> ast_combine(
+    const tdb_unique_ptr<ASTNode>& lhs,
+    const tdb_unique_ptr<ASTNode>& rhs,
+    QueryConditionCombinationOp combination_op) {
+  // AST Construction.
+  std::vector<tdb_unique_ptr<ASTNode>> ast_nodes;
+  if (lhs->get_tag() == ASTNodeTag::VAL) {
```

Comment:
This check is repeated a lot of times. To make the code more readable you could create a `vritual` member method in your node classes to do such checks and remove the need for a tag, e.g.:

```
virtual bool ASTNode::is_expr() = 0;

bool ASTNodeExpr::is_expr() override {
  return true;
}

bool ASTNodeVal::is_expr() override {
  return false;
}
```

Reply:
Hmm.  Is `get_tag()` already a virtual function?  Then having just a function call to `is_expr()` instead of function call to `get_val()` plus comparison will be *slightly* faster – but the function call overhead is going to overwhelm the cost of the comparison so the difference might not be noticeable.

Taking a step back I am wondering how much explicit checking of tags (whether via `get_tag()` or via `is_expr()`) you really need to be doing at this level.  Per Ypatia's comments below, you *should* be able to get most or all of this functionality via virtual functions.  It seems like all that is happening is you need to decide whether to push back the child's nodes as your own, or whether to just push back the child.  (See comment below, which thanks to me not pushing the right button, may have actually arrived earlier.)

Reply:
We still need to do some, since the behavior can change in the QC class where we run the apply_trees and stuff.